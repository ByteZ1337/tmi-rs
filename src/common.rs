use std::cell::RefCell;

/// Channel name known to be prefixed by `#`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Channel<'src>(&'src str);

impl<'src> Channel<'src> {
  pub fn as_str(&self) -> &'src str {
    self.0
  }

  pub fn parse(s: &'src str) -> Result<Self, InvalidChannelName> {
    match s.starts_with('#') {
      true => Ok(Self(s)),
      false => Err(InvalidChannelName),
    }
  }

  pub(crate) fn from_unchecked(s: &'src str) -> Self {
    Self(s)
  }
}

impl<'src> std::fmt::Display for Channel<'src> {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.write_str(self.0)
  }
}

impl<'src> TryFrom<&'src str> for Channel<'src> {
  type Error = InvalidChannelName;

  fn try_from(value: &'src str) -> Result<Self, Self::Error> {
    Self::parse(value)
  }
}

#[derive(Debug)]
pub struct InvalidChannelName;
impl std::fmt::Display for InvalidChannelName {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.write_str("channel name is missing \"#\" prefix")
  }
}
impl std::error::Error for InvalidChannelName {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Span {
  pub start: u32,
  pub end: u32,
}

impl Span {
  #[allow(dead_code)]
  #[doc(hidden)]
  #[inline]
  pub(crate) fn get<'src>(&self, src: &'src str) -> &'src str {
    &src[*self]
  }
}

impl From<std::ops::Range<usize>> for Span {
  #[inline]
  fn from(value: std::ops::Range<usize>) -> Self {
    Span {
      start: value.start as u32,
      end: value.end as u32,
    }
  }
}

impl From<Span> for std::ops::Range<usize> {
  #[inline]
  fn from(value: Span) -> Self {
    value.start as usize..value.end as usize
  }
}

impl std::ops::Index<Span> for str {
  type Output = <str as std::ops::Index<std::ops::Range<usize>>>::Output;

  #[inline]
  fn index(&self, index: Span) -> &Self::Output {
    self.index(std::ops::Range::from(index))
  }
}

pub struct Join<I, S>(RefCell<Option<I>>, S);

impl<I, S> std::fmt::Display for Join<I, S>
where
  // TODO: get rid of this `Clone` bound by doing `peek`
  // manually
  I: Iterator,
  <I as Iterator>::Item: std::fmt::Display,
  S: std::fmt::Display,
{
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    let Some(iter) = self.0.borrow_mut().take() else {
      return Err(std::fmt::Error);
    };

    let sep = &self.1;
    let mut peekable = iter.peekable();
    while let Some(item) = peekable.next() {
      write!(f, "{item}")?;
      if peekable.peek().is_some() {
        write!(f, "{sep}")?;
      }
    }
    Ok(())
  }
}

pub trait JoinIter: Sized {
  fn join<Sep>(self, sep: Sep) -> Join<Self, Sep>;
}

impl<Iter> JoinIter for Iter
where
  Iter: Sized + Iterator,
{
  fn join<Sep>(self, sep: Sep) -> Join<Self, Sep> {
    Join(RefCell::new(Some(self)), sep)
  }
}

#[doc(hidden)]
pub mod unescaped;
