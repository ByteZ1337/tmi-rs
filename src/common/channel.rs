use std::borrow::Borrow;
use std::ops::Deref;

/// Channel name known to be prefixed by `#`.
#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(transparent)]
pub struct ChannelRef(str);

impl ChannelRef {
  /// Get the string value of the channel name.
  pub fn as_str(&self) -> &str {
    &self.0
  }

  /// Parse a string into a channel name.
  ///
  /// The channel name must begin with a `#` character.
  pub fn parse(s: &str) -> Result<&Self, InvalidChannelName> {
    match s.starts_with('#') {
      true => Ok(Self::from_unchecked(s)),
      false => Err(InvalidChannelName),
    }
  }

  pub(crate) fn from_unchecked(s: &str) -> &Self {
    // # Safety:
    // - `Self` is `repr(transparent)` and only holds a single `str` field,
    //   therefore the layout of `Self` is the same as `str`, and it's
    //   safe to transmute between the two
    unsafe { std::mem::transmute(s) }
  }
}

impl Deref for ChannelRef {
  type Target = str;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl AsRef<str> for ChannelRef {
  fn as_ref(&self) -> &str {
    &self.0
  }
}

impl AsRef<ChannelRef> for ChannelRef {
  fn as_ref(&self) -> &ChannelRef {
    self
  }
}

impl Borrow<str> for ChannelRef {
  fn borrow(&self) -> &str {
    &self.0
  }
}

impl std::fmt::Debug for ChannelRef {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.debug_tuple("Channel").field(&self.as_str()).finish()
  }
}

impl std::fmt::Display for ChannelRef {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.write_str(&self.0)
  }
}

impl ToOwned for ChannelRef {
  type Owned = Channel;

  fn to_owned(&self) -> Self::Owned {
    Channel::from_unchecked(self.as_str().to_owned())
  }
}

/// Channel name known to be prefixed by `#`.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(transparent)]
pub struct Channel(String);

impl Channel {
  /// Get the string value of the channel name.
  pub fn as_str(&self) -> &str {
    self.0.as_str()
  }

  /// Parse a string into a channel name.
  ///
  /// The channel name must begin with a `#` character.
  pub fn parse(s: String) -> Result<Self, InvalidChannelName> {
    match s.starts_with('#') {
      true => Ok(Self(s)),
      false => Err(InvalidChannelName),
    }
  }

  pub(crate) fn from_unchecked(s: String) -> Self {
    Self(s)
  }
}

impl Deref for Channel {
  type Target = String;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl AsRef<str> for Channel {
  fn as_ref(&self) -> &str {
    self.0.as_ref()
  }
}

impl AsRef<ChannelRef> for Channel {
  fn as_ref(&self) -> &ChannelRef {
    ChannelRef::from_unchecked(self.0.as_str())
  }
}

impl Borrow<str> for Channel {
  fn borrow(&self) -> &str {
    self.0.borrow()
  }
}

impl Borrow<ChannelRef> for Channel {
  fn borrow(&self) -> &ChannelRef {
    ChannelRef::from_unchecked(self.0.borrow())
  }
}

impl std::fmt::Display for Channel {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.write_str(&self.0)
  }
}

/// Failed to parse a channel name.
#[derive(Debug)]
pub struct InvalidChannelName;
impl std::fmt::Display for InvalidChannelName {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.write_str("channel name is missing \"#\" prefix")
  }
}
impl std::error::Error for InvalidChannelName {}

static_assert_send!(ChannelRef);
static_assert_sync!(ChannelRef);

static_assert_send!(Channel);
static_assert_sync!(Channel);
