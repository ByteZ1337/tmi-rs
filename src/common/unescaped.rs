use beef::lean::Cow;
use std::cell::Cell;
use std::cell::UnsafeCell;

pub struct Unescaped<'a> {
  escaped: Cell<bool>,
  inner: UnsafeCell<Cow<'a, str>>,
}

assert_send!(Unescaped);

impl<'a> Unescaped<'a> {
  pub fn new(s: impl Into<Cow<'a, str>>) -> Self {
    Self {
      escaped: Cell::new(false),
      inner: UnsafeCell::new(s.into()),
    }
  }

  #[inline]
  pub fn get(&self) -> &str {
    if !self.escaped.get() {
      self.unescape();
    }

    unsafe { (*self.inner.get()).as_ref() }
  }

  #[cold]
  fn unescape(&self) {
    unsafe { unescape_in_place(&mut *self.inner.get()) }
    self.escaped.set(true);
  }
}

fn unescape_in_place(value: &mut Cow<'_, str>) {
  #[inline]
  fn actually_unescape(input: &str, start: usize) -> String {
    let mut out = String::with_capacity(input.len());
    out.push_str(&input[..start]);

    let mut escape = false;
    for char in input[start..].chars() {
      match char {
        '\\' if escape => {
          out.push('\\');
          escape = false;
        }
        '\\' => escape = true,
        ':' if escape => {
          out.push(';');
          escape = false;
        }
        's' if escape => {
          out.push(' ');
          escape = false;
        }
        'r' if escape => {
          out.push('\r');
          escape = false;
        }
        'n' if escape => {
          out.push('\n');
          escape = false;
        }
        '⸝' => out.push(','),
        c => out.push(c),
      }
    }

    out
  }

  for i in 0..value.len() {
    if value.as_bytes()[i] == b'\\' {
      *value = Cow::owned(actually_unescape(value, i));
      break;
    }
  }
}

impl<'a> From<&'a str> for Unescaped<'a> {
  fn from(value: &'a str) -> Self {
    Self::new(value)
  }
}

impl<'a> Clone for Unescaped<'a> {
  fn clone(&self) -> Self {
    Self {
      escaped: self.escaped.clone(),
      inner: UnsafeCell::new(unsafe { (*self.inner.get()).clone() }),
    }
  }
}

impl<'a> std::fmt::Debug for Unescaped<'a> {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    std::fmt::Debug::fmt(self.get(), f)
  }
}

impl<'a> std::hash::Hash for Unescaped<'a> {
  fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
    self.get().hash(state)
  }
}
impl<'a> std::cmp::PartialEq for Unescaped<'a> {
  fn eq(&self, other: &Self) -> bool {
    self.get() == other.get()
  }
}
impl<'a> std::cmp::Eq for Unescaped<'a> {}
impl<'a> std::cmp::PartialOrd for Unescaped<'a> {
  fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
    Some(self.cmp(other))
  }
}
impl<'a> std::cmp::Ord for Unescaped<'a> {
  fn cmp(&self, other: &Self) -> std::cmp::Ordering {
    self.get().cmp(other.get())
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn unescape_in_place_non_escaped() {
    let mut v = "test".into();
    unescape_in_place(&mut v);
    assert!(v.is_borrowed());
    assert_eq!(v, "test");
  }

  #[test]
  fn unescape_in_place_escaped() {
    let mut v = "\\\\\\n\\s".into();
    unescape_in_place(&mut v);
    assert!(v.is_owned());
    assert_eq!(v, "\\\n ");
  }

  #[test]
  fn unescape_system_msg() {
    let v = Unescaped::new("An\\sanonymous\\suser\\sgifted\\sa\\sTier\\s1\\ssub\\sto\\sDot0422!");
    assert_eq!(v.get(), "An anonymous user gifted a Tier 1 sub to Dot0422!");
  }
}
