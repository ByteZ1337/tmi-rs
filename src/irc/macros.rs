#[doc(hidden)]
#[macro_export]
macro_rules! __count {
  () => (0usize);
  ($x:tt $($xs:tt)*) => (1usize + $crate::__count!($($xs)*));
}

#[macro_export]
macro_rules! whitelist {
  [$($tag:ident),*] => (
    unsafe {
      $crate::irc::Whitelist::<{$crate::__count!($($tag)*)}, _>::new({
        #[allow(unused_variables)]
        #[inline]
        |src: &str, map: &mut $crate::irc::RawTags, tag: $crate::common::Span, value: $crate::common::Span| {
          match src[tag].as_bytes() {
            $($crate::irc::tags::$tag => {map.push($crate::irc::RawTagPair($crate::irc::RawTag::$tag, value));})*
            _ => {}
          };
        }
      })
    }
  )
}
