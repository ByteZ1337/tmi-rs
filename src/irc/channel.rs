use crate::common::Span;

/// #channel <rest>
#[inline(always)]
pub(super) fn parse(src: &str, pos: &mut usize) -> Option<Span> {
  match src[*pos..].starts_with('#') {
    true => {
      let start = *pos;
      match src[start..].find(' ') {
        Some(end) => {
          let end = start + end;
          *pos = end + 1;
          Some(Span::from(start..end))
        }
        None => {
          let end = src.len();
          *pos = end;
          Some(Span::from(start..end))
        }
      }
    }
    false => None,
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn channel() {
    let data = "#channel <rest>";
    let mut pos = 0;

    let channel = parse(data, &mut pos).unwrap();
    assert_eq!(channel.get(data), "#channel");
    assert_eq!(&data[pos..], "<rest>");
  }
}
