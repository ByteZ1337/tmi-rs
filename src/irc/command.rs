use std::fmt::Display;

use crate::common::Span;

#[derive(Clone, Copy)]
pub(super) enum RawCommand {
  Ping,
  Pong,
  Join,
  Part,
  Privmsg,
  Whisper,
  Clearchat,
  Clearmsg,
  GlobalUserState,
  Notice,
  Reconnect,
  RoomState,
  UserNotice,
  UserState,
  Capability,
  RplWelcome,
  RplYourHost,
  RplCreated,
  RplMyInfo,
  RplNamReply,
  RplEndOfNames,
  RplMotd,
  RplMotdStart,
  RplEndOfMotd,
  Other(Span),
}

impl RawCommand {
  #[inline]
  pub(super) fn get<'src>(&self, src: &'src str) -> Command<'src> {
    match self {
      RawCommand::Ping => Command::Ping,
      RawCommand::Pong => Command::Pong,
      RawCommand::Join => Command::Join,
      RawCommand::Part => Command::Part,
      RawCommand::Privmsg => Command::Privmsg,
      RawCommand::Whisper => Command::Whisper,
      RawCommand::Clearchat => Command::ClearChat,
      RawCommand::Clearmsg => Command::ClearMsg,
      RawCommand::GlobalUserState => Command::GlobalUserState,
      RawCommand::Notice => Command::Notice,
      RawCommand::Reconnect => Command::Reconnect,
      RawCommand::RoomState => Command::RoomState,
      RawCommand::UserNotice => Command::UserNotice,
      RawCommand::UserState => Command::UserState,
      RawCommand::Capability => Command::Capability,
      RawCommand::RplWelcome => Command::RplWelcome,
      RawCommand::RplYourHost => Command::RplYourHost,
      RawCommand::RplCreated => Command::RplCreated,
      RawCommand::RplMyInfo => Command::RplMyInfo,
      RawCommand::RplNamReply => Command::RplNames,
      RawCommand::RplEndOfNames => Command::RplEndOfNames,
      RawCommand::RplMotd => Command::RplMotd,
      RawCommand::RplMotdStart => Command::RplMotdStart,
      RawCommand::RplEndOfMotd => Command::RplEndOfMotd,
      RawCommand::Other(span) => Command::Other(&src[*span]),
    }
  }
}

/// A Twitch IRC command.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Command<'src> {
  /// Ping the peer
  Ping,
  /// The peer's response to a [`Command::Ping`]
  Pong,
  /// Join a channel
  Join,
  /// Leave a channel
  Part,
  /// Send a message to a channel
  Privmsg,
  /// Send a private message to a user
  Whisper,
  /// Purge a user's messages in a channel
  ClearChat,
  /// Remove a single message
  ClearMsg,
  /// Sent upon successful authentication (PASS/NICK command)
  GlobalUserState,
  /// General notices from the server
  Notice,
  /// Rejoins channels after a restart
  Reconnect,
  /// Identifies the channel's chat settings
  RoomState,
  /// Announces Twitch-specific events to the channel
  UserNotice,
  /// Identifies a user's chat settings or properties
  UserState,
  /// Requesting an IRC capability
  Capability,
  // Numeric commands
  /// `001`
  RplWelcome,
  /// `002`
  RplYourHost,
  /// `003`
  RplCreated,
  /// `004`
  RplMyInfo,
  /// `353`
  RplNames,
  /// `366`
  RplEndOfNames,
  /// `372`
  RplMotd,
  /// `375`
  RplMotdStart,
  /// `376`
  RplEndOfMotd,
  /// Unknown command
  Other(&'src str),
}

impl<'src> Display for Command<'src> {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.write_str(self.as_str())
  }
}

impl<'src> Command<'src> {
  /// Get the string value of the [`Command`].
  pub fn as_str(&self) -> &'src str {
    use Command::*;
    match self {
      Ping => "PING",
      Pong => "PONG",
      Join => "JOIN",
      Part => "PART",
      Privmsg => "PRIVMSG",
      Whisper => "WHISPER",
      ClearChat => "CLEARCHAT",
      ClearMsg => "CLEARMSG",
      GlobalUserState => "GLOBALUSERSTATE",
      Notice => "NOTICE",
      Reconnect => "RECONNECT",
      RoomState => "ROOMSTATE",
      UserNotice => "USERNOTICE",
      UserState => "USERSTATE",
      Capability => "CAP",
      RplWelcome => "001",
      RplYourHost => "002",
      RplCreated => "003",
      RplMyInfo => "004",
      RplNames => "353",
      RplEndOfNames => "366",
      RplMotd => "372",
      RplMotdStart => "375",
      RplEndOfMotd => "376",
      Other(cmd) => cmd,
    }
  }
}

/// `COMMAND <rest>`
///
/// Returns `None` if command is unknown *and* empty
#[inline(always)]
pub(super) fn parse(src: &str, pos: &mut usize) -> Option<RawCommand> {
  let (end, next_pos) = match src[*pos..].find(' ') {
    Some(end) => {
      let end = *pos + end;
      (end, end + 1)
    }
    None => (src.len(), src.len()),
  };

  use RawCommand as C;
  let cmd = match &src[*pos..end] {
    "PING" => C::Ping,
    "PONG" => C::Pong,
    "JOIN" => C::Join,
    "PART" => C::Part,
    "PRIVMSG" => C::Privmsg,
    "WHISPER" => C::Whisper,
    "CLEARCHAT" => C::Clearchat,
    "CLEARMSG" => C::Clearmsg,
    "GLOBALUSERSTATE" => C::GlobalUserState,
    "NOTICE" => C::Notice,
    "RECONNECT" => C::Reconnect,
    "ROOMSTATE" => C::RoomState,
    "USERNOTICE" => C::UserNotice,
    "USERSTATE" => C::UserState,
    "CAP" => C::Capability,
    "001" => C::RplWelcome,
    "002" => C::RplYourHost,
    "003" => C::RplCreated,
    "004" => C::RplMyInfo,
    "353" => C::RplNamReply,
    "366" => C::RplEndOfNames,
    "372" => C::RplMotd,
    "375" => C::RplMotdStart,
    "376" => C::RplEndOfMotd,
    other if !other.is_empty() => C::Other(Span::from(*pos..end)),
    _ => return None,
  };

  *pos = next_pos;

  Some(cmd)
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn command() {
    let data = "PING <rest>";
    let mut pos = 0;

    let command = parse(data, &mut pos).unwrap();
    assert_eq!(command.get(data), Command::Ping);
    assert_eq!(&data[pos..], "<rest>");
  }
}
