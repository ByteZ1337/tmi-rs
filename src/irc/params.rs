use crate::common::Span;

#[inline(always)]
pub(super) fn parse(src: &str, pos: &usize) -> Option<Span> {
  if !src[*pos..].is_empty() {
    Some(Span::from(*pos..src.len()))
  } else {
    None
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn params() {
    let data = ":param_a :param_b";
    let params = parse(data, &0).unwrap();
    assert_eq!(params.get(data), data)
  }
}
