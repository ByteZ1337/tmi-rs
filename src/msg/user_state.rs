//! Sent upon joining a channel, or upon successfully sending a `PRIVMSG` message to a channel.
//!
//! This is like [`GlobalUserState`][crate::msg::global_user_state::GlobalUserState], but
//! carries channel-specific information.
//!
//! For example, [`UserState::badges`] may be different from [`GlobalUserState::badges`][crate::msg::global_user_state::GlobalUserState::badges].

use super::{is_not_empty, parse_badges, split_comma, Badge, MessageParseError};
use crate::common::Channel;
use crate::irc::{Command, IrcMessageRef, Tag};

/// Sent upon joining a channel, or upon successfully sending a `PRIVMSG` message to a channel.
///
/// This is like [`GlobalUserState`][crate::msg::global_user_state::GlobalUserState], but
/// carries channel-specific information.
///
/// For example, [`UserState::badges`] may be different from [`GlobalUserState::badges`][crate::msg::global_user_state::GlobalUserState::badges].
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct UserState<'src> {
  channel: Channel<'src>,
  user_name: &'src str,
  badges: Vec<Badge<'src>>,
  emote_sets: Vec<&'src str>,
  color: Option<&'src str>,
}

generate_getters! {
  <'src> for UserState<'src> as self {
    /// Name of the channel in which this state applies to.
    channel -> Channel<'_>,

    /// Display name of the user.
    user_name -> &str,

    /// List of channel-specific badges.
    badges -> &[Badge<'_>] = self.badges.as_ref(),

    /// Emote sets which are available in this channel.
    emote_sets -> &[&str] = self.emote_sets.as_ref(),

    /// The user's selected name color.
    ///
    /// [`None`] means the user has not selected a color.
    /// To match the behavior of Twitch, users should be
    /// given a globally-consistent random color.
    color -> Option<&str>,
  }
}

impl<'src> UserState<'src> {
  fn parse(message: IrcMessageRef<'src>) -> Option<Self> {
    if message.command() != Command::UserState {
      return None;
    }

    Some(UserState {
      channel: message.channel()?,
      user_name: message.tag(Tag::DisplayName)?,
      badges: message
        .tag(Tag::Badges)
        .zip(message.tag(Tag::BadgeInfo))
        .map(|(badges, badge_info)| parse_badges(badges, badge_info))
        .unwrap_or_default(),
      emote_sets: message
        .tag(Tag::EmoteSets)
        .map(split_comma)
        .map(Iterator::collect)
        .unwrap_or_default(),
      color: message.tag(Tag::Color).filter(is_not_empty),
    })
  }
}

impl<'src> super::FromIrc<'src> for UserState<'src> {
  #[inline]
  fn from_irc(message: IrcMessageRef<'src>) -> Result<Self, MessageParseError> {
    Self::parse(message).ok_or(MessageParseError)
  }
}

impl<'src> From<UserState<'src>> for super::Message<'src> {
  fn from(msg: UserState<'src>) -> Self {
    super::Message::UserState(msg)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn parse_userstate() {
    assert_irc_snapshot!(UserState, "@badge-info=;badges=;color=#FF0000;display-name=TESTUSER;emote-sets=0;mod=0;subscriber=0;user-type= :tmi.twitch.tv USERSTATE #randers");
  }

  #[test]
  fn parse_userstate_uuid_emote_set_id() {
    assert_irc_snapshot!(UserState, "@badge-info=;badges=moderator/1;color=#8A2BE2;display-name=TESTUSER;emote-sets=0,75c09c7b-332a-43ec-8be8-1d4571706155;mod=1;subscriber=0;user-type=mod :tmi.twitch.tv USERSTATE #randers");
  }
}
