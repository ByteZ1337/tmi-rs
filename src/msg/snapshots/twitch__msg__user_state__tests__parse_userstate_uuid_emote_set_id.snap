---
source: src/msg/user_state.rs
expression: "f(\"@badge-info=;badges=moderator/1;color=#8A2BE2;display-name=TESTUSER;emote-sets=0,75c09c7b-332a-43ec-8be8-1d4571706155;mod=1;subscriber=0;user-type=mod :tmi.twitch.tv USERSTATE #randers\")"
---
UserState {
    channel: Channel(
        "#randers",
    ),
    user_name: "TESTUSER",
    badges: [
        Moderator,
    ],
    emote_sets: [
        "0",
        "75c09c7b-332a-43ec-8be8-1d4571706155",
    ],
    color: Some(
        "#8A2BE2",
    ),
}
